src/main/java/net/neoforged/neoforge/transfer/ResourceHandlerUtil.java

```diff
@@ -0,0 +1,565 @@
+            int size = handler.size();
+            for (int index = 0; index < size; index++) {
+                extracted += handler.extract(index, resource, amount - extracted, subTransaction);
+                if (extracted >= amount)
+                    return extracted;
+            }
+            subTransaction.commit();
+            return extracted;
+        }
+    }
+
+    public static <T extends IResource> int getTotalAmountOf(IResourceHandler<T> handler, T resource) {
+        try (Transaction transaction = TransactionManager.open(TransactionContext.ROOT)) {
+            //We don't commit allow us to just inquiry the amount
+            return extract(handler, resource, Integer.MAX_VALUE, transaction);
+        }
+    }
+
+    /**
+     * Extracts the first resource from an {@link IResourceHandler} that matches the given filter.
+     *
+     * @param <R>          The type of resource handled by the handler
+     * @param <S>          The type of stack created by the extraction
+     * @param handler      The {@link IResourceHandler} to extract the resource from
+     * @param filter       The filter to apply to the resources
+     * @param amount       The desired amount of the resource to extract
+     * @param transaction  The transaction context for a given insertion.
+     *                     Passing in {@code null} will essentially be the same as doing `execute`,
+     *                     whereas passing in a closeable context allows you to choose if it should be committed.
+     * @param stackFactory A factory the given a resource of type {@code <R>} and an amount, a stack of type {@code <S>} can be created. The return is expected to be non-null and properly be the instanced empty value for a given resource.
+     * @return a stack of type {@code <S>} typically in the form of an ResourceStack or as an example an ItemStack based on the factory provided
+     */
+    public static <R extends IResource, S> S extractFiltered(
+            IResourceHandler<R> handler,
+            Predicate<R> filter,
+            int amount,
+            R defaultResource,
+            @Nullable TransactionContext transaction,
+            IStackFactory<R, S> stackFactory) {
+        try (Transaction subTransaction = TransactionManager.open(transaction)) {
+            int size = handler.size();
+            int handled = 0;
+            R resourceTarget = defaultResource;
+            for (int index = 0; index < size; index++) {
+                R resource = handler.getResource(index);
+                if (resource.isEmpty()) continue;
+                if (!filter.test(resource)) continue;
+                if (resourceTarget.isEmpty())
+                    resourceTarget = resource;
+                else if (!resourceTarget.equals(resource)) continue;
+
+                handled += handler.extract(resource, amount - handled, subTransaction);
+                if (handled == amount) {
+                    subTransaction.commit();
```

Comment:
Rather than remembering to have to manually declare commit here, we could replace this and the return with a break statement.